### PR TITLE
Add real tests for Queue Managers and fix a number of issues

### PR DESCRIPTION
**Pull request by Lnaden**
## Description

In addressing #287 and #288, I figured out how to get some real tests in on the Queue Managers. They can only test that the `adapter` and `QueueManager` constructs are built correctly, so they cannot submit jobs and therefore cannot properly test running on Schedulers like SLURM, PBS, SGE, etc. That said, its still a start and adding the tests revealed a few bugs. Itemized list of changes below:

* Make auto-doc generator work for Pydantic 0.28. Will work with pre 0.28 as well
* Removed the Manager Template Generator 
* Added real tests for the Queue Managers
* Tests revealed a number of corner cases which needed fixing
* Fixed example doc for Manager
* Made a default Dask/Cluster/Parsl set for Queue Managers which don't
  set them (or worse, set them and leave them empty). Potential
  resolution to #288

## Questions
- [x] @jeffschriber this PR does make default constructs for the dask/parsl/cluster blocks and should also avoid unhelpful error from #288. I'm not sure this is quite everything you wanted from that issue (i.e. more clarity on what items need set when in blocks like `parsl`), but in case it is, let me know if we can close that issue. If not, lets leave it open and keep working through that.

## Status
- [x] Changelog updated
- [x] Ready to go

### CODE REVIEW

File: qcfractal/cli/qcfractal_manager.py [hunk 1]
```diff
@@ -439,9 +441,9 @@ class ManagerSettings(BaseModel):
     common: CommonManagerSettings = CommonManagerSettings()
     server: FractalServerSettings = FractalServerSettings()
     manager: QueueManagerSettings = QueueManagerSettings()
-    cluster: Optional[ClusterSettings] = None
-    dask: Optional[DaskQueueSettings] = None
-    parsl: Optional[ParslQueueSettings] = None
+    cluster: Optional[ClusterSettings] = ClusterSettings()
+    dask: Optional[DaskQueueSettings] = DaskQueueSettings()
+    parsl: Optional[ParslQueueSettings] = ParslQueueSettings()
```
Comment (dgasmith) [inline qcfractal/cli/qcfractal_manager.py]: +1 for this. No reason not to auto init I think.

File: qcfractal/cli/tests/test_cli.py [hunk 2]
```diff
@@ -70,28 +72,54 @@ def test_manager_executor_manager_boot_from_file(active_server, tmp_path):
...
-    if adapter == "parsl" and scheduler == "lsf":
-        pytest.xfail("Parsl has no LSF implementation")
-
-    tmpl_path = tmp_path / "tmp_template.py"
-    args = ["qcfractal-template", adapter, scheduler, "--test", "-o", str(tmpl_path)]
-    testing.run_process(args)
+def cli_manager_runs(config_data, tmp_path):
+    temp_config = tmp_path / "temp_config.yaml"
+    with open(temp_config, 'w') as config:
+        config.write(yaml.dump(config_data))
```
Comment (dgasmith) [inline qcfractal/cli/tests/test_cli.py]: Hmm, might be better to use the `temp_config.write(yaml.dumps())`?

File: qcfractal/cli/tests/test_cli.py [hunk 3]
```diff
@@ -70,28 +72,54 @@ def test_manager_executor_manager_boot_from_file(active_server, tmp_path):
...
+    ("dask", "lSf"),
+    ("parsl", "slurm"),
+    ("parsl", "PBS"),
+    ("parsl", "MoAb"),
+    ("parsl", "SGE"),
+    pytest.param("parsl", "lSf", marks=pytest.mark.xfail),
+    pytest.param("NotAParser", "slurm", marks=pytest.mark.xfail),
+    pytest.param("dask", "NotAScheduler", marks=pytest.mark.xfail),
+])
+def test_cli_managers(adapter, scheduler, tmp_path):
```
Comment (dgasmith) [inline qcfractal/cli/tests/test_cli.py]: How much does this hit the CI?

File: qcfractal/cli/tests/test_cli.py [hunk 4]
```diff
@@ -70,28 +72,54 @@ def test_manager_executor_manager_boot_from_file(active_server, tmp_path):
...
+    if config_present is not None:
+        config[adapter] = None
+
+
+def test_cli_managers_quick_exits():
+    """Test that --help and --schema correctly work"""
+    args = ["qcfractal-manager", "--help"]
+    testing.run_process(args)
+    args = ["qcfractal-manager", "--schema"]
+    testing.run_process(args)
```
Comment (dgasmith) [inline qcfractal/cli/tests/test_cli.py]: Two separate tests perhaps and this should be:
```python
_options = {"coverage": True, "dump_stdout": True}
testing.run_process(args, interupt_after=10, **_options)
```

Same with the above, we should keep all CLI tests somewhat equal.

File: qcfractal/cli/tests/test_cli.py (hunk 3 above)
Comment (Lnaden) [reply qcfractal/cli/tests/test_cli.py]: Do you mean time wise how much does each of these take to run on the CI?

Comment (dgasmith) [reply qcfractal/cli/tests/test_cli.py]: Yes.

Comment (Lnaden) [reply qcfractal/cli/tests/test_cli.py]: Each one takes anywhere from 3-10 seconds. And much of that is a function of which adapter is being used

Comment (dgasmith) [reply qcfractal/cli/tests/test_cli.py]: Gotcha, seems reasonable.

**Comment by jeffschriber** [comment]
Adding the default inits definitely will solve the issue I was having, thanks! Enabling these minimal config files should also be helpful for general usability (at least by non dask/parsl experts like myself). The docs do touch on some useful options, though I was a bit thrown off by the dask options used in the example YAML files not playing well with the PBS queuing system (e.g. `queue: default`). I think a big part of my confusion was just a general unfamiliarity with dask/parsl which isn't the responsibility of the fractal docs.

**Comment by Lnaden** [comment]
I finalized the last test I meant to do and then reworked them a bit to reduce the time, the whole set of 19 tests takes ~85 seconds now, It was ~155s before the re-work.

I think this is ready to go now

File: qcfractal/cli/tests/test_cli.py [hunk 5]
```diff
@@ -70,28 +71,79 @@ def test_manager_executor_manager_boot_from_file(active_server, tmp_path):
...
+def load_manager_config(adapter, scheduler):
+    config = read_config_file(os.path.join(_pwd, "manager_boot_template.yaml"))
+    config["common"]["adapter"] = adapter
+    config["cluster"]["scheduler"] = scheduler
+    return config
+
+
+@testing.mark_slow
+@testing.using_dask_jobqueue
+@testing.using_parsl
```
Comment (dgasmith) [inline qcfractal/cli/tests/test_cli.py]: One last thing here is that it requires both Parsl and Dask to be present. I think you can do:

```
testing.using_dask_jobqueue("dask", "slurm"),
```

If not can you remove the above decorators and within the test_cli_managers function use a 
```
if (adapter == "dask") and not testing.has_module("Dask"):
    pytest.mark.skip("reasons)
```

**Comment by Lnaden** [comment]
It turns out it was not quite that, but close.

`pytest.param("dask", "slurm", marks=testing.using_dask_jobqueue)`

And to further my love of pytest, we can multi-mark as well, *because of course we can,* apparently!

`pytest.param("parsl", "lSf", marks=[testing.using_parsl, pytest.mark.xfail])`
